fundingmanager.go

```diff
@@ -2580,7 +2650,44 @@ func (f *fundingManager) handleErrorMsg(fmsg *fundingErrorMsg) {
 	}
 }
 
-// cancelReservationCtx do all needed work in order to securely cancel the
+// handlePruneMsg processes a prune message sent by the zombie sweeper
+// by looping through each pending reservation and cancelling reservations
+// that are considered zombies (have not been updated since ReservationTimeout).
+// A zero value on the last update timestamp represents a reservation that is
+// waiting for the funding transaction. Those reservations are not cancelled.
+func (f *fundingManager) handlePruneMsg(msg *pruneMsg) {
+	f.resMtx.Lock()
+	defer f.resMtx.Unlock()
+
+	for peerIDKey, pendingReservations := range f.activeReservations {
+		for pendingChanID, resCtx := range pendingReservations {
+			if !resCtx.lastUpdated.IsZero() && time.Since(resCtx.lastUpdated) > f.cfg.ReservationTimeout {
+				errMsg := &lnwire.Error{
+					ChanID: pendingChanID,
+					Data:   lnwire.ErrorData("funding failed due to internal error"),
```

Comment:
The reason we can't call `failFundingFlow` here is because a few lines above we acquire a write lock on the `resMtx` so that our access to the map is threadsafe. `failFundingFlow` eventually calls `deleteReservationCtx` which attempts to acquire a write lock on `resMtx`. Thus, if we called `failFundingFlow` within this loop we would end up in deadlock. Even if we switched the above lock to a read lock we would end up in deadlock because `deleteReservationCtx` requires a write lock.

Reply:
this can be fixed by fetching all the reservations to be deleted under the read lock, then calling `failFundingFlow` on them all after releasing it.